[#66411] score input for item form component

Kharonus opened the pull request:
# Ticket
[#66411](https://community.openproject.org/work_packages/66411)

# What are you trying to accomplish?
- render form differently depending on custom field format

Kharonus committed:
[#66411] score input for item form component

- https://community.openproject.org/work_packages/66411
- render form differently depending on custom field format

Kharonus commented:
Waiting for #19973 to get merged first.

🟢 Done

NobodysNightmare commented on app/forms/custom_fields/hierarchy/item_form.rb:
```diff
@@ -46,16 +46,30 @@ class ItemForm < ApplicationForm
-            placeholder: I18n.t("custom_fields.admin.items.placeholder.short"),
-            validation_message: validation_message_for(:short)
-          )
+          if @input_format == :short
```
:green_circle: I'd have a slight preference for a case statement here, because we switch between a number of known values for `input_format`.

NobodysNightmare replied:
:yellow_circle: After re-reading: Does it make sense to not if-else this, but rather use the input format on the components of the input?

Hopefully complete example of my idea:

```ruby
input_group.text_field(
  name: @input_format,
  label: I18n.t("does.not.yet.exist.#{@input_format}"),
  type: @input_format == :score ? :number : :text,
  value: @target_item.public_send(@input_format),
  visually_hide_label: true,
  full_width: false,
  required: true,
  placeholder: I18n.t("custom_fields.admin.items.placeholder.#{@input_format}"),
  validation_message: validation_message_for(@input_format)
)
```

mereghost replied:
Either this or break the block into something along the lines of:

```ruby
# in the block
secondary_input(input_group)

# later on the class
def secondary_input(input_group)
  return short_input(input_group) if @input_format == :short

  score_input(input_group)
end

Kharonus replied:
I'm on marcello's point here. I'd rather leave conditional logic out of the field, but construct well separated input definitions. extracted those into methods and used switch.

NobodysNightmare commented on app/forms/custom_fields/hierarchy/item_form.rb:
```diff
@@ -46,16 +46,30 @@ class ItemForm < ApplicationForm
+          if @input_format == :short
+            input_group.text_field(
+              name: :short,
+              label: "Short",
```
:red_circle: I18n seems to be missing here and below for the score.

edit: And apparently above for the "Label" as well... I can confirm this locally. The thing is that the Label is only rendered for screenreaders, but that doesn't make it less important.

Kharonus replied:
very good hint. will add the localized text here

NobodysNightmare commented on app/components/admin/custom_fields/hierarchy/item_form_component.rb:
```diff
@@ -57,6 +57,18 @@ def url
           end
         end
 
+        def input_format
```
:yellow_circle: I'd suggest double thinking about the name of this method.

The `ItemFormComponent` has two inputs:

* The one for the label (the big one)
* The one for the score or the short (the small one)

This format here only defines the small one, not the big one.

Brainstorming:

* extra_input_format
* second_input_format

mereghost replied:
`secondary_field_format` also sounds good.

Kharonus replied:
chose `secondary_field_format`

NobodysNightmare reviewed (changes requested):
I found one blocking issue. Generally looks good to me.

Kharonus committed:
[#66411] made hidden labels localised

- important for screen readers
- used better method and variable naming
- extracted methods for better readability

Kharonus committed:
[#66411] updated ui test targets
